% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/align_target.R
\name{combined_header}
\alias{combined_header}
\title{Create a combined .bam header}
\usage{
combined_header(bam_files, header_file = "header_tmp.sam")
}
\arguments{
\item{bam_files}{A list of the locations/file names of .bam files from which
to combine the headers.}

\item{header_file}{A file name and location for the output file for the
combined header. This will be a .sam format file without any reads.
Defaults to 'header_tmp.sam'.}
}
\value{
This function will return a combined header from all the supplied .bam files.
}
\description{
This function generates a combined header from multiple .bam files from
different reference libraries (e.g. a split bacterial library).
It is not intended for use by users.
}
\examples{

refPath <- system.file("extdata","Mononegavirales.fasta", package = "MetaScope")
file.copy(from = refPath, to = file.path(".", "Mononegavirales.fasta"))
mk_subread_index('Mononegavirales.fasta', split = .002)

readPath <- system.file("extdata", "virus_example.fastq", package = "MetaScope")
Rsubread::align(index = "Mononegavirales_1", readfile1 = readPath,
output_file = "virus_example1.bam")
Rsubread::align(index = "Mononegavirales_2", readfile1 = readPath,
output_file = "virus_example2.bam")

bam_files <- c('virus_example1.bam','virus_example2.bam')
com_head <- combined_header(bam_files)

}
